File: AnkOutput/android_female_cycle/data/FemaleCycleDataManager.kt
package com.yourapp.health.female.data

import android.content.Context
import android.content.SharedPreferences
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.yourapp.health.female.data.model.CycleConfiguration
import com.yourapp.health.female.data.model.DailySymptomData
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.HashMap

/**
 * 女性生理周期数据管理器(单例)
 * 对应iOS端的FemaleCycleDataManager类
 *
 * 功能:
 * - 管理周期配置(经期天数、周期长度、最后经期日期)
 * - 管理每日症状数据
 * - 本地数据持久化
 * - 数据变更通知机制
 * - 周期计算(经期、排卵期、预测经期)
 */
class FemaleCycleDataManager private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var INSTANCE: FemaleCycleDataManager? = null

        fun getInstance(context: Context): FemaleCycleDataManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: FemaleCycleDataManager(context.applicationContext).also { INSTANCE = it }
            }
        }

        // 存储键
        private const val PREFS_NAME = "FemaleCycle_Data"
        private const val KEY_CYCLE_CONFIG = "FemaleCycle_Configuration"
        private const val KEY_DAILY_DATA = "FemaleCycle_DailyData"
    }

    // SharedPreferences
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val gson = Gson()

    // 日期格式化器
    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

    // 周期配置
    private var cycleConfig: CycleConfiguration = CycleConfiguration()

    // 每日症状数据字典，key为日期字符串（yyyy-MM-dd）
    private val dailyDataDict: HashMap<String, DailySymptomData> = HashMap()

    // LiveData for data change notifications
    private val _cycleConfigChanged = MutableLiveData<CycleConfiguration>()
    val cycleConfigChanged: LiveData<CycleConfiguration> = _cycleConfigChanged

    private val _dailyDataChanged = MutableLiveData<Pair<String, DailySymptomData>>()
    val dailyDataChanged: LiveData<Pair<String, DailySymptomData>> = _dailyDataChanged

    // Alias for symptomDataChanged to match iOS naming
    private val _symptomDataChanged = MutableLiveData<String>()
    val symptomDataChanged: LiveData<String> = _symptomDataChanged

    init {
        loadAllData()
    }

    // ==================== 数据加载 ====================

    /**
     * 从本地存储加载所有数据
     */
    private fun loadAllData() {
        // 加载周期配置
        val configJson = prefs.getString(KEY_CYCLE_CONFIG, null)
        cycleConfig = if (configJson != null) {
            try {
                gson.fromJson(configJson, CycleConfiguration::class.java)
            } catch (e: Exception) {
                CycleConfiguration()
            }
        } else {
            // 兼容旧数据(如果有)
            CycleConfiguration(
                periodDays = prefs.getInt("FemaleHealth_PeriodDays", 7),
                cycleLength = prefs.getInt("FemaleHealth_CycleLength", 28),
                lastPeriodDate = prefs.getLong("FemaleHealth_LastPeriodDate", System.currentTimeMillis())
            )
        }

        // 加载每日数据
        val dailyDataJson = prefs.getString(KEY_DAILY_DATA, null)
        dailyDataDict.clear()
        if (dailyDataJson != null) {
            try {
                val type = object : TypeToken<HashMap<String, DailySymptomData>>() {}.type
                val loadedData: HashMap<String, DailySymptomData> = gson.fromJson(dailyDataJson, type)
                dailyDataDict.putAll(loadedData)
            } catch (e: Exception) {
                // 解析失败,使用空数据
            }
        }
    }

    /**
     * 保存所有数据到本地存储
     */
    private fun saveAllData() {
        prefs.edit().apply {
            putString(KEY_CYCLE_CONFIG, gson.toJson(cycleConfig))
            putString(KEY_DAILY_DATA, gson.toJson(dailyDataDict))
        }.apply()
    }

    // ==================== 周期配置管理 ====================

    /**
     * 更新周期配置
     */
    fun updateCycleConfiguration(
        periodDays: Int? = null,
        cycleLength: Int? = null,
        lastPeriodDate: Long? = null
    ) {
        var hasChanges = false

        periodDays?.let {
            if (it != cycleConfig.periodDays && it in 3..10) {
                cycleConfig.periodDays = it
                hasChanges = true
            }
        }

        cycleLength?.let {
            if (it != cycleConfig.cycleLength && it in 21..35) {
                cycleConfig.cycleLength = it
                hasChanges = true
            }
        }

        lastPeriodDate?.let {
            // 比较日期是否为同一天
            val calendar = Calendar.getInstance()
            calendar.timeInMillis = it
            calendar.set(Calendar.HOUR_OF_DAY, 0)
            calendar.set(Calendar.MINUTE, 0)
            calendar.set(Calendar.SECOND, 0)
            calendar.set(Calendar.MILLISECOND, 0)
            val newDate = calendar.timeInMillis

            calendar.timeInMillis = cycleConfig.lastPeriodDate
            calendar.set(Calendar.HOUR_OF_DAY, 0)
            calendar.set(Calendar.MINUTE, 0)
            calendar.set(Calendar.SECOND, 0)
            calendar.set(Calendar.MILLISECOND, 0)
            val oldDate = calendar.timeInMillis

            if (newDate != oldDate) {
                cycleConfig.lastPeriodDate = newDate
                hasChanges = true
            }
        }

        if (hasChanges) {
            cycleConfig.isConfigured = true
            saveAllData()
            _cycleConfigChanged.postValue(cycleConfig)
        }
    }

    /**
     * 获取周期配置
     */
    fun getCycleConfiguration(): CycleConfiguration {
        return cycleConfig.copy()
    }

    /**
     * 检查是否已配置
     */
    fun isConfigured(): Boolean {
        return cycleConfig.isConfigured && cycleConfig.isValid()
    }

    // ==================== 每日数据管理 ====================

    /**
     * 获取指定日期的症状数据
     */
    fun getDailyData(date: Long): DailySymptomData {
        val dateString = dateFormatter.format(Date(date))
        return dailyDataDict[dateString] ?: DailySymptomData(dateString)
    }

    /**
     * 获取指定日期字符串的症状数据
     */
    fun getDailyData(dateString: String): DailySymptomData {
        return dailyDataDict[dateString] ?: DailySymptomData(dateString)
    }

    /**
     * 更新指定日期的症状数据
     */
    fun updateDailyData(date: Long, data: DailySymptomData) {
        val dateString = dateFormatter.format(Date(date))
        dailyDataDict[dateString] = data
        saveAllData()
        _dailyDataChanged.postValue(Pair(dateString, data))
        _symptomDataChanged.postValue(dateString)
    }

    /**
     * 更新指定日期的经期开始状态
     */
    fun updatePeriodStartStatus(date: Long, isStarted: Boolean) {
        val data = getDailyData(date).copy(isPeriodStarted = isStarted)
        updateDailyData(date, data)
    }

    /**
     * 更新指定日期的流量等级
     */
    fun updateFlowLevel(date: Long, level: Int) {
        val data = getDailyData(date).copy(flowLevel = level)
        updateDailyData(date, data)
    }

    /**
     * 更新指定日期的痛经等级
     */
    fun updatePainLevel(date: Long, level: Int) {
        val data = getDailyData(date).copy(painLevel = level)
        updateDailyData(date, data)
    }

    /**
     * 更新指定日期的性行为
     */
    fun updateSexualActivity(date: Long, activity: Int) {
        val data = getDailyData(date).copy(sexualActivity = activity)
        updateDailyData(date, data)
    }

    /**
     * 更新指定日期的心情
     */
    fun updateMood(date: Long, mood: Int) {
        val data = getDailyData(date).copy(mood = mood)
        updateDailyData(date, data)
    }

    /**
     * 更新指定日期的身体症状
     */
    fun updateBodySymptoms(date: Long, symptoms: List<String>) {
        val data = getDailyData(date)
        data.bodySymptoms.clear()
        data.bodySymptoms.addAll(symptoms)
        updateDailyData(date, data)
    }

    /**
     * 检查指定日期是否有记录数据
     */
    fun hasRecordData(date: Long): Boolean {
        val dateString = dateFormatter.format(Date(date))
        return dailyDataDict.containsKey(dateString)
    }

    /**
     * 获取所有有记录的日期
     */
    fun getAllRecordedDates(): List<String> {
        return dailyDataDict.keys.toList()
    }

    /**
     * 获取所有有记录的日期（按时间倒序排列）
     */
    fun getAllRecordedDatesSorted(): List<String> {
        return dailyDataDict.keys.sortedDescending()
    }

    /**
     * 按月份分组获取所有记录（倒序）
     */
    fun getRecordsByMonth(): List<Pair<String, List<String>>> {
        val sortedDates = getAllRecordedDatesSorted()
        val result = mutableListOf<Pair<String, List<String>>>()
        var currentMonth = ""
        var currentDates = mutableListOf<String>()

        for (dateString in sortedDates) {
            // 提取月份："yyyy-MM"
            val monthString = dateString.substring(0, 7) // "2025-12-09" -> "2025-12"

            if (monthString != currentMonth) {
                if (currentDates.isNotEmpty()) {
                    result.add(Pair(currentMonth, currentDates.toList()))
                }
                currentMonth = monthString
                currentDates = mutableListOf(dateString)
            } else {
                currentDates.add(dateString)
            }
        }

        // 添加最后一组
        if (currentDates.isNotEmpty()) {
            result.add(Pair(currentMonth, currentDates.toList()))
        }

        return result
    }

    /**
     * 获取所有记录（转换为DailyRecord列表）
     * 用于AllDataActivity显示
     */
    fun getAllRecords(): List<com.yourapp.health.female.data.model.DailyRecord> {
        val records = mutableListOf<com.yourapp.health.female.data.model.DailyRecord>()

        for ((dateString, symptomData) in dailyDataDict) {
            try {
                val date = dateFormatter.parse(dateString)
                val timestamp = date?.time ?: continue

                // 计算周期天数
                val cycleDay = calculateCycleDay(timestamp)

                // 创建DailyRecord
                val record = com.yourapp.health.female.data.model.DailyRecord(
                    date = dateString,
                    cycleDay = cycleDay,
                    isPeriod = cycleDay > 0,
                    hasFlow = symptomData.flowLevel > 0,
                    hasPain = symptomData.painLevel > 0,
                    hasTemp = false, // 暂未支持体温
                    hasSexual = symptomData.sexualActivity > 0,
                    hasMood = symptomData.mood > 0,
                    hasBodySymptoms = symptomData.bodySymptoms.isNotEmpty(),
                    recordTime = timestamp
                )
                records.add(record)
            } catch (e: Exception) {
                // 跳过无效日期
            }
        }

        // 按日期倒序排列
        return records.sortedByDescending { it.date }
    }

    /**
     * 检查指定日期是否有症状记录
     */
    fun hasSymptomRecord(dateString: String): Boolean {
        val symptomData = dailyDataDict[dateString] ?: return false

        // 检查是否有任何症状数据
        return symptomData.flowLevel > 0 ||
               symptomData.painLevel > 0 ||
               symptomData.sexualActivity > 0 ||
               symptomData.mood > 0 ||
               symptomData.bodySymptoms.isNotEmpty() ||
               symptomData.isPeriodStarted
    }

    // ==================== 别名方法（兼容iOS命名） ====================

    /**
     * 获取症状数据（别名方法，匹配iOS命名）
     */
    fun getSymptomData(dateString: String): DailySymptomData {
        return getDailyData(dateString)
    }

    /**
     * 保存症状数据（别名方法，匹配iOS命名）
     */
    fun saveSymptomData(data: DailySymptomData) {
        try {
            val date = dateFormatter.parse(data.date)
            if (date != null) {
                updateDailyData(date.time, data)
            }
        } catch (e: Exception) {
            // 解析失败，直接保存
            dailyDataDict[data.date] = data
            saveAllData()
            _dailyDataChanged.postValue(Pair(data.date, data))
            _symptomDataChanged.postValue(data.date)
        }
    }

    /**
     * 获取周期天数（别名方法，匹配iOS命名）
     */
    fun getCycleDayNumber(date: Long): Int {
        return calculateCycleDay(date)
    }

    // ==================== 周期计算 ====================

    /**
     * 计算指定日期在周期中的天数（1表示经期第一天）
     * 返回0表示不在经期内
     */
    fun calculateCycleDay(date: Long): Int {
        val calendar = Calendar.getInstance()
        val lastPeriodCalendar = Calendar.getInstance()
        lastPeriodCalendar.timeInMillis = cycleConfig.lastPeriodDate

        // 计算与上次经期开始的天数差
        val days = ((date - cycleConfig.lastPeriodDate) / (24 * 60 * 60 * 1000)).toInt()

        // 如果是负数，说明在上次经期之前
        if (days < 0) {
            return 0
        }

        // 检查是否在当前周期的经期内
        if (days < cycleConfig.periodDays) {
            return days + 1 // 经期第1-7天
        }

        // 检查是否在下一个周期的经期内
        val nextPeriodStart = cycleConfig.lastPeriodDate + (cycleConfig.cycleLength * 24 * 60 * 60 * 1000L)
        val daysFromNext = ((date - nextPeriodStart) / (24 * 60 * 60 * 1000)).toInt()

        if (daysFromNext >= 0 && daysFromNext < cycleConfig.periodDays) {
            return daysFromNext + 1
        }

        return 0 // 不在经期内
    }

    /**
     * 判断指定日期是否在经期内
     */
    fun isInPeriod(date: Long): Boolean {
        return calculateCycleDay(date) > 0
    }

    /**
     * 计算经期日期集合
     */
    fun getPeriodDates(centerDate: Long): Set<String> {
        val dates = mutableSetOf<String>()
        val calendar = Calendar.getInstance()

        // 计算当前经期
        for (i in 0 until cycleConfig.periodDays) {
            calendar.timeInMillis = cycleConfig.lastPeriodDate
            calendar.add(Calendar.DAY_OF_YEAR, i)
            dates.add(dateFormatter.format(calendar.time))
        }

        // 计算下一个经期(如果在显示范围内)
        calendar.timeInMillis = cycleConfig.lastPeriodDate
        calendar.add(Calendar.DAY_OF_YEAR, cycleConfig.cycleLength)
        val nextPeriodStart = calendar.timeInMillis

        for (i in 0 until cycleConfig.periodDays) {
            calendar.timeInMillis = nextPeriodStart
            calendar.add(Calendar.DAY_OF_YEAR, i)
            dates.add(dateFormatter.format(calendar.time))
        }

        return dates
    }

    /**
     * 计算排卵期日期集合
     * 排卵日 = 下次月经第一天 - 14天
     * 排卵期 = 排卵日前5天到排卵日后4天，共10天
     */
    fun getOvulationDates(centerDate: Long): Set<String> {
        val dates = mutableSetOf<String>()
        val calendar = Calendar.getInstance()

        val ovulationDayOffset = cycleConfig.cycleLength - 14
        val ovulationStart = ovulationDayOffset - 5
        val ovulationEnd = ovulationDayOffset + 4

        // 获取经期日期(避免重复)
        val periodDates = getPeriodDates(centerDate)

        for (i in ovulationStart..ovulationEnd) {
            calendar.timeInMillis = cycleConfig.lastPeriodDate
            calendar.add(Calendar.DAY_OF_YEAR, i)
            val dateString = dateFormatter.format(calendar.time)

            // 只有不在经期的日期才加入排卵期集合
            if (!periodDates.contains(dateString)) {
                dates.add(dateString)
            }
        }

        return dates
    }

    /**
     * 计算预测经期日期集合
     */
    fun getPredictedPeriodDates(centerDate: Long): Set<String> {
        val dates = mutableSetOf<String>()
        val calendar = Calendar.getInstance()

        // 计算下一次经期
        calendar.timeInMillis = cycleConfig.lastPeriodDate
        calendar.add(Calendar.DAY_OF_YEAR, cycleConfig.cycleLength)
        val nextPeriodStart = calendar.timeInMillis

        for (i in 0 until cycleConfig.periodDays) {
            calendar.timeInMillis = nextPeriodStart
            calendar.add(Calendar.DAY_OF_YEAR, i)
            dates.add(dateFormatter.format(calendar.time))
        }

        return dates
    }

    // ==================== 数据清理 ====================

    /**
     * 清除指定日期之前的旧数据（保留最近N天）
     */
    fun cleanOldData(keepDays: Int = 365) {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.DAY_OF_YEAR, -keepDays)
        val cutoffDateString = dateFormatter.format(calendar.time)

        var removedCount = 0
        val iterator = dailyDataDict.iterator()
        while (iterator.hasNext()) {
            val entry = iterator.next()
            if (entry.key < cutoffDateString) {
                iterator.remove()
                removedCount++
            }
        }

        if (removedCount > 0) {
            saveAllData()
        }
    }

    /**
     * 清除所有数据（谨慎使用）
     */
    fun clearAllData() {
        dailyDataDict.clear()
        cycleConfig = CycleConfiguration()
        saveAllData()

        _cycleConfigChanged.postValue(cycleConfig)
        _dailyDataChanged.postValue(Pair("", DailySymptomData("")))
    }
}

File: SmartBracelet/Health/FemaleCycleDataManager.swift
//
// Copyright © 2015-2018 bruce Innovations Technology Limited All Rights Reserved.
// The program and materials is not free. Without our permission, any use, including but not limited to reproduction, retransmission, communication, display, mirror, download, modification, is expressly prohibited. Otherwise, it will be pursued for legal liability.
//
//  FemaleCycleDataManager.swift
//  SmartBracelet
//
//  Created by bruce on 2025/12/11.
//  Copyright © 2025 tjd. All rights reserved.
//

import Foundation

// MARK: - 日期症状数据模型

struct DailySymptomData: Codable {
    var date: String // 格式: yyyy-MM-dd
    var isPeriodStarted: Bool
    var flowLevel: Int // 0=未选择, 1=少, 2=中, 3=多
    var painLevel: Int // 0=未选择, 1=轻微, 2=中等, 3=严重
    var sexualActivity: Int // 0=无, 1=保护性行为, 2=无保护性行为
    var mood: Int // 0=未选择, 1=平静, 2=开心, 3=放松, 4=活力满满, 5=敏感, 6=焦躁, 7=易怒, 8=悲伤
    var bodySymptoms: [String] // 身体症状列表（格式：分类-症状）

    init(date: String) {
        self.date = date
        self.isPeriodStarted = false
        self.flowLevel = 0
        self.painLevel = 0
        self.sexualActivity = 0
        self.mood = 0
        self.bodySymptoms = []
    }
}

// MARK: - 周期配置数据模型

struct CycleConfiguration: Codable {
    var periodDays: Int // 经期天数
    var cycleLength: Int // 周期长度
    var lastPeriodDate: Date // 最后一次经期开始日期
    var isConfigured: Bool // 是否已由用户配置过

    init(periodDays: Int = 7, cycleLength: Int = 28, lastPeriodDate: Date = Date(), isConfigured: Bool = false) {
        self.periodDays = periodDays
        self.cycleLength = cycleLength
        self.lastPeriodDate = lastPeriodDate
        self.isConfigured = isConfigured
    }
}

// MARK: - 数据变更通知

extension Notification.Name {
    static let femaleCycleDataDidChange = Notification.Name("FemaleCycleDataDidChange")
    static let cycleConfigurationDidChange = Notification.Name("CycleConfigurationDidChange")
}

// MARK: - 生理周期数据管理器

class FemaleCycleDataManager {

    // MARK: - 单例

    static let shared = FemaleCycleDataManager()

    private init() {
        loadAllData()
    }

    // MARK: - 属性

    // 周期配置
    private(set) var cycleConfig: CycleConfiguration = CycleConfiguration()

    // 每日症状数据字典，key为日期字符串（yyyy-MM-dd）
    private var dailyDataDict: [String: DailySymptomData] = [:]

    // 日期格式化器
    private let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    // MARK: - 存储键

    private let cycleConfigKey = "FemaleCycle_Configuration"
    private let dailyDataKey = "FemaleCycle_DailyData"

    // MARK: - 数据加载

    private func loadAllData() {
        // 加载周期配置
        if let data = UserDefaults.standard.data(forKey: cycleConfigKey),
           let config = try? JSONDecoder().decode(CycleConfiguration.self, from: data) {
            cycleConfig = config
        } else {
            // 兼容旧数据
            let periodDays = UserDefaults.standard.integer(forKey: "FemaleHealth_PeriodDays")
            let cycleLength = UserDefaults.standard.integer(forKey: "FemaleHealth_CycleLength")
            let lastPeriodDateTimestamp = UserDefaults.standard.double(forKey: "FemaleHealth_LastPeriodDate")

            cycleConfig = CycleConfiguration(
                periodDays: periodDays > 0 ? periodDays : 7,
                cycleLength: cycleLength > 0 ? cycleLength : 28,
                lastPeriodDate: lastPeriodDateTimestamp > 0 ? Date(timeIntervalSince1970: lastPeriodDateTimestamp) : Date()
            )
        }

        // 加载每日数据
        if let data = UserDefaults.standard.data(forKey: dailyDataKey),
           let dict = try? JSONDecoder().decode([String: DailySymptomData].self, from: data) {
            dailyDataDict = dict
        }
    }

    // MARK: - 数据保存

    private func saveAllData() {
        // 保存周期配置
        if let data = try? JSONEncoder().encode(cycleConfig) {
            UserDefaults.standard.set(data, forKey: cycleConfigKey)
        }

        // 保存每日数据
        if let data = try? JSONEncoder().encode(dailyDataDict) {
            UserDefaults.standard.set(data, forKey: dailyDataKey)
        }

        // 同步到磁盘
        UserDefaults.standard.synchronize()

        XLogger.shared.log("数据已保存 - 周期配置: \(cycleConfig), 每日数据条数: \(dailyDataDict.count)")
    }

    // MARK: - 周期配置管理

    /// 更新周期配置
    func updateCycleConfiguration(periodDays: Int? = nil, cycleLength: Int? = nil, lastPeriodDate: Date? = nil) {
        var hasChanges = false

        if let periodDays = periodDays, periodDays != cycleConfig.periodDays {
            cycleConfig.periodDays = periodDays
            hasChanges = true
        }

        if let cycleLength = cycleLength, cycleLength != cycleConfig.cycleLength {
            cycleConfig.cycleLength = cycleLength
            hasChanges = true
        }

        if let lastPeriodDate = lastPeriodDate {
            let calendar = Calendar.current
            if !calendar.isDate(lastPeriodDate, inSameDayAs: cycleConfig.lastPeriodDate) {
                cycleConfig.lastPeriodDate = lastPeriodDate
                hasChanges = true
            }
        }

        if hasChanges {
            // 标记为已配置
            cycleConfig.isConfigured = true

            saveAllData()

            // 发送配置变更通知
            NotificationCenter.default.post(name: .cycleConfigurationDidChange, object: nil)

            XLogger.shared.log("周期配置已更新: periodDays=\(cycleConfig.periodDays), cycleLength=\(cycleConfig.cycleLength), lastPeriodDate=\(cycleConfig.lastPeriodDate), isConfigured=\(cycleConfig.isConfigured)")
        }
    }

    /// 获取周期配置
    func getCycleConfiguration() -> CycleConfiguration {
        return cycleConfig
    }

    // MARK: - 每日数据管理

    /// 获取指定日期的症状数据
    func getDailyData(for date: Date) -> DailySymptomData {
        let dateString = dateFormatter.string(from: date)
        return dailyDataDict[dateString] ?? DailySymptomData(date: dateString)
    }

    /// 获取指定日期字符串的症状数据
    func getDailyData(for dateString: String) -> DailySymptomData {
        return dailyDataDict[dateString] ?? DailySymptomData(date: dateString)
    }

    /// 更新指定日期的症状数据
    func updateDailyData(for date: Date, data: DailySymptomData) {
        let dateString = dateFormatter.string(from: date)
        dailyDataDict[dateString] = data
        saveAllData()

        // 发送数据变更通知
        NotificationCenter.default.post(
            name: .femaleCycleDataDidChange,
            object: nil,
            userInfo: ["date": dateString, "data": data]
        )

        XLogger.shared.log("每日数据已更新: \(dateString) - \(data)")
    }

    /// 更新指定日期的经期开始状态
    func updatePeriodStartStatus(for date: Date, isStarted: Bool) {
        var data = getDailyData(for: date)
        data.isPeriodStarted = isStarted
        updateDailyData(for: date, data: data)
    }

    /// 更新指定日期的流量等级
    func updateFlowLevel(for date: Date, level: Int) {
        var data = getDailyData(for: date)
        data.flowLevel = level
        updateDailyData(for: date, data: data)
    }

    /// 更新指定日期的痛经等级
    func updatePainLevel(for date: Date, level: Int) {
        var data = getDailyData(for: date)
        data.painLevel = level
        updateDailyData(for: date, data: data)
    }

    /// 更新指定日期的性行为
    func updateSexualActivity(for date: Date, activity: Int) {
        var data = getDailyData(for: date)
        data.sexualActivity = activity
        updateDailyData(for: date, data: data)
    }

    /// 更新指定日期的心情
    func updateMood(for date: Date, mood: Int) {
        var data = getDailyData(for: date)
        data.mood = mood
        updateDailyData(for: date, data: data)
    }

    /// 更新指定日期的身体症状
    func updateBodySymptoms(for date: Date, symptoms: [String]) {
        var data = getDailyData(for: date)
        data.bodySymptoms = symptoms
        updateDailyData(for: date, data: data)
    }

    /// 检查指定日期是否有记录数据
    func hasRecordData(for date: Date) -> Bool {
        let dateString = dateFormatter.string(from: date)
        // 只要该日期在字典中存在，就算有记录
        return dailyDataDict[dateString] != nil
    }

    /// 获取所有有记录的日期
    func getAllRecordedDates() -> [String] {
        return Array(dailyDataDict.keys)
    }

    /// 获取所有有记录的日期（按时间倒序排列）
    func getAllRecordedDatesSorted() -> [String] {
        return dailyDataDict.keys.sorted(by: >)
    }

    /// 按月份分组获取所有记录（倒序）
    func getRecordsByMonth() -> [(month: String, dates: [String])] {
        let sortedDates = getAllRecordedDatesSorted()
        var result: [(month: String, dates: [String])] = []
        var currentMonth = ""
        var currentDates: [String] = []

        for dateString in sortedDates {
            // 提取月份："yyyy-MM"
            let monthString = String(dateString.prefix(7)) // "2025-12-09" -> "2025-12"

            if monthString != currentMonth {
                if !currentDates.isEmpty {
                    result.append((month: currentMonth, dates: currentDates))
                }
                currentMonth = monthString
                currentDates = [dateString]
            } else {
                currentDates.append(dateString)
            }
        }

        // 添加最后一组
        if !currentDates.isEmpty {
            result.append((month: currentMonth, dates: currentDates))
        }

        return result
    }

    /// 计算指定日期在周期中的天数（1表示经期第一天）
    /// 返回0表示不在经期内
    func calculateCycleDay(for date: Date) -> Int {
        let calendar = Calendar.current
        let lastPeriodDate = cycleConfig.lastPeriodDate

        // 计算与上次经期开始的天数差
        let days = calendar.dateComponents([.day], from: lastPeriodDate, to: date).day ?? 0

        // 如果是负数，说明在上次经期之前
        if days < 0 {
            return 0
        }

        // 检查是否在当前周期的经期内
        if days < cycleConfig.periodDays {
            return days + 1 // 经期第1-7天
        }

        // 检查是否在下一个周期的经期内
        let nextPeriodStart = calendar.date(byAdding: .day, value: cycleConfig.cycleLength, to: lastPeriodDate)
        if let nextStart = nextPeriodStart {
            let daysFromNext = calendar.dateComponents([.day], from: nextStart, to: date).day ?? -1
            if daysFromNext >= 0 && daysFromNext < cycleConfig.periodDays {
                return daysFromNext + 1
            }
        }

        return 0 // 不在经期内
    }

    /// 判断指定日期是否在经期内
    func isInPeriod(date: Date) -> Bool {
        return calculateCycleDay(for: date) > 0
    }

    // MARK: - 数据清理

    /// 清除指定日期之前的旧数据（保留最近N天）
    func cleanOldData(keepDays: Int = 365) {
        let calendar = Calendar.current
        let cutoffDate = calendar.date(byAdding: .day, value: -keepDays, to: Date()) ?? Date()
        let cutoffDateString = dateFormatter.string(from: cutoffDate)

        var removedCount = 0
        for (dateString, _) in dailyDataDict {
            if dateString < cutoffDateString {
                dailyDataDict.removeValue(forKey: dateString)
                removedCount += 1
            }
        }

        if removedCount > 0 {
            saveAllData()
            XLogger.shared.log("清理了 \(removedCount) 条旧数据")
        }
    }

    /// 清除所有数据（谨慎使用）
    func clearAllData() {
        dailyDataDict.removeAll()
        cycleConfig = CycleConfiguration()
        saveAllData()

        // 发送通知
        NotificationCenter.default.post(name: .cycleConfigurationDidChange, object: nil)
        NotificationCenter.default.post(name: .femaleCycleDataDidChange, object: nil)

        XLogger.shared.log("所有数据已清除")
    }
}
